drivers/cc2420/cc2420.c

```diff
@@ -216,32 +217,43 @@ void cc2420_tx_exec(cc2420_t *dev)
 int cc2420_rx(cc2420_t *dev, uint8_t *buf, size_t max_len, void *info)
 {
     uint8_t len;
+    uint8_t crc_corr;
 
-    /* get the packet length (without dropping it) (first byte in RX FIFO */
+    /* get the packet length (without dropping it) (first byte in RX FIFO) */
     cc2420_ram_read(dev, CC2420_RAM_RXFIFO, &len, 1);
+
+    /* fetch and check if CRC_OK bit (MSB) is set */
+    cc2420_ram_read(dev, (CC2420_RAM_RXFIFO + len - 1), &crc_corr, 1);
+    if (!(crc_corr & 0x80)) {
+        DEBUG("cc2420: recv: CRC_OK bit not set, dropping packet\n");
+        /* drop the corrupted frame from the RXFIFO */
+        cc2420_strobe(dev, CC2420_STROBE_FLUSHRX);
+        cc2420_strobe(dev, CC2420_STROBE_FLUSHRX);
+        return 0;
+    }
+
     len -= 2;   /* subtract RSSI and FCF */
 
     if (!buf) {
         DEBUG("cc2420: recv: packet of length %i in RX FIFO\n", (int)len);
+        return 0;
     }
 
     /* if a buffer is given, read (and drop) the packet */
-    if (buf) {
-        len = (len > max_len) ? max_len : len;
-
-        /* drop length byte */
-        cc2420_fifo_read(dev, NULL, 1);
-        /* read fifo contents */
-        DEBUG("cc2420: recv: reading %i byte of the packet\n", (int)len);
-        cc2420_fifo_read(dev, buf, len);
-
-        uint8_t rssi;
-        cc2420_fifo_read(dev, &rssi, 1);
-        DEBUG("cc2420: recv: RSSI is %i\n", (int)rssi);
-        /* finally flush the FIFO */
-        cc2420_strobe(dev, CC2420_STROBE_FLUSHRX);
-        cc2420_strobe(dev, CC2420_STROBE_FLUSHRX);
-    }
+    len = (len > max_len) ? max_len : len;
```

Comment:
Now that you compared to the Atmel driver, it [returns an error](https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_netdev.c#L154) in case of a packet that is too long. However, IMO that must not be addressed inside this PR. I already opened an other PR addressing several improvements for the cc2420 which is still WIP.